distributed/process.py

```diff
@@ -81,15 +81,26 @@ def __init__(self, loop=None, target=None, name=None, args=(), kwargs={}):
                 dask.config.global_config,
             ),
         )
-        _dangling.add(self._process)
         self._name = self._process.name
+        self._proc_finalizer = weakref.finalize(self, self._get_finalizer())
```

Comment:
It's important that a finalizer contain no references to the object being finalized, otherwise a cycle will prevent it from ever being run.

Right now, `_get_finalizer` does the right thing, but IMO it'd be clearer to define the finalizer as an external function and pass in the `process` as an arg, making it crystal clear that the finalizer contains no references to `self`.

Reply:
> It's important that a finalizer contain no references to the object being finalized, otherwise a cycle will prevent it from ever being run.

Sorry for the syntactical annoyance, a cycle _will_ or _may_ prevent it? I'm asking because then we may need to fix most of the uses we have of it in distributed today, including in `AsyncProcess` itself, a few examples:

https://github.com/dask/distributed/blob/18fff8b0ca0297e100695519931c78e579cb55a6/distributed/process.py#L121

https://github.com/dask/distributed/blob/18fff8b0ca0297e100695519931c78e579cb55a6/distributed/comm/tcp.py#L143

https://github.com/dask/distributed/blob/18fff8b0ca0297e100695519931c78e579cb55a6/distributed/scheduler.py#L1512

It's not super difficult to change that, so I can probably work on a separate PR for that if you think that has a big potential for problems.

Reply:
> Right now, `_get_finalizer` does the right thing, but IMO it'd be clearer to define the finalizer as an external function and pass in the `process` as an arg, making it crystal clear that the finalizer contains no references to `self`.

Fixed in https://github.com/dask/distributed/pull/4184/commits/d61f92932a9aa775d436f05125f5096a8cd2cf1a

Reply:
> a cycle will or may prevent it

*Will* prevent it. Finalizers that contain references to `obj` will never run automatically (manually calling them will still work), as the cycle will prevent automatic collection.

https://docs.python.org/3/library/weakref.html#weakref.finalize

To be more specific, given the signature `finalize(obj, func, *args, **kwargs)`, none of `func`, `args`, or `kwargs` can contain a reference to `obj`. The examples you posted above are all fine.

Reply:
> The examples you posted above are all fine as they don't break this rule.

Were you referring to `self._process` that I had in `finalize(proc=self._process)`? If so, then we do have that issue in the TCP example I pasted above if you look at the finalizer:

https://github.com/dask/distributed/blob/18fff8b0ca0297e100695519931c78e579cb55a6/distributed/comm/tcp.py#L164-L170

Reply:
That code is fine since no where in `def finalize` is `self` bound as part of the closure. If we were writing that code today, I'd pull that function out to a top level function and pass in the arguments as `args`/`kwargs to `weakref.finalize` (as I asked you to do here), but since it already exists and is correct (if a bit opaque) I wouldn't mess with it.